### PR TITLE
Minor fixes to sparse setindex!

### LINKED ISSUE #16216: Bug in sparse matrices
We have that this works

R = spzeros(4,5)
R[:,6] = spzeros(4,1)

I believe that the code should return an error similar to
R = zeros(4,5)
R[:,6] = zeros(4,1)

ERROR: BoundsError: attempt to access 4x5 Array{Float64,2}:

### PR DESCRIPTION
**Pull request by martinholters**

### CODE REVIEW

**Comment by martinholters** [comment]
Ah crap, missed the possibility of indexing with an empty vector. This was not tested for, but sure enough bit me in my own project.

**Comment by martinholters** [comment]
Huh, interestingly, for a sparse matrix `A`, in master we have `A[1:0,:] = []` working (being a no-op), while `A[:,1:0] = []` errors. I will include a fix to this in this PR.

**Comment by martinholters** [comment]
Ok, indexing with empty vectors now returns before doing any real work (thus not erroring), but after checking the other index is within bounds to throw a `BoundsError` even for things like `A[1:0, -1]`. This seems to be consistent with `Array` behavior.

**Comment by martinholters** [comment]
Argh, this looked so innocent at first... Of course, `setindex!` also dispatches on what is being assigned (single element or matrix), so this needs to be fixed at more places. Stay tuned...

**Comment by martinholters** [comment]
I hope I caught all cases this time.

**Comment by andreasnoack** [comment]
cc: @mbauman

**Comment by mbauman** [comment]
:+1:

It'd be really nice to just punt to `checkbounds` here, but of course this is much more efficient since we know the array is sorted.  Makes me want for a SortedVector type.  But that's for another day.

File: base/sparse/sparsematrix.jl [hunk 1]
```diff
@@ -2265,7 +2265,15 @@ function spset!{Tv,Ti<:Integer}(A::SparseMatrixCSC{Tv}, x::Tv, I::AbstractVector
...
     m, n = size(A)
     lenI = length(I)
-    ((I[end] > m) || (J[end] > n)) && throw(DimensionMismatch(""))
+
+    if (!isempty(I) && (I[1] < 1 || I[end] > m)) || (!isempty(J) && (J[1] < 1 || J[end] > n))
+        throw(BoundsError(A, (I, J)))
+    end
+
+    if isempty(I) || isempty(J)
+        return A
```
Comment (tkelman) [inline base/sparse/sparsematrix.jl]: shouldn't these be returning the rhs, not the lhs?

Comment (mbauman) [inline base/sparse/sparsematrix.jl]: Returning the RHS is something the parser does on its own:

```jl
julia> expand(:(x[i] = 1))
:(begin
        setindex!(x,1,i)
        return 1
    end)
```

`setindex!` and most `mutating!` functions have an unwritten (?) convention to return the object that gets modified.

**Comment by StefanKarpinski** [comment]
> Makes me want for a SortedVector type.

This would make so many APIs so much nicer – and there would be so many lovely, easy speedups.

**Comment by johnmyleswhite** [comment]
Seems cheap to add once you've got stack allocation of non-pointer-free immutables

File: base/sparse/sparsematrix.jl (hunk 1 above)
Comment (martinholters) [inline base/sparse/sparsematrix.jl]: And the functions in question here return the LHS at the end of their normal code path, so I thought I better stick to that.

**Comment by martinholters** [comment]
Anyone got an idea about the travis failure? (No, it's not an OSX timeout.) A segfault in `typemap.c` while testing docs - looks unrelated, but certainly disconcerting.

**Comment by martinholters** [comment]
While at it, fixed another issue: We had `A[1:2] = 1:3` working in master (by just assigning 1 and 2). Now throws an appropriate `DimensionMismatch`.

File: base/sparse/sparsematrix.jl [hunk 2]
```diff
@@ -2671,14 +2693,19 @@ function setindex!{Tv,Ti,T<:Real}(A::SparseMatrixCSC{Tv,Ti}, x, I::AbstractVecto
     S = issorted(I) ? (1:n) : sortperm(I)
     sxidx = r1 = r2 = 0
 
+    if (!isempty(I) && (I[S[1]] < 1 || I[S[end]] > length(A)))
+        throw(BoundsError(A, I))
+    end
+
+    isa(x, AbstractArray) && setindex_shape_check(x, length(I))
```
Comment (tkelman) [inline base/sparse/sparsematrix.jl]: is `setindex_shape_check` documented somewhere?

Comment (martinholters) [inline base/sparse/sparsematrix.jl]: Not that I'm aware of.

**Comment by martinholters** [comment]
Ah, the Travis error is gone. In case someone would like to take a look at it, [here is the log](https://travis-ci.org/JuliaLang/julia/jobs/128829733).

**Comment by timholy** [comment]
@vtjnash, do most of [these lines](https://github.com/JuliaLang/julia/blob/89487db348d43ab1de369a7ff8b73196689d1265/src/typemap.c#L857-L862) need a `gb_wb`?

**Comment by yuyichao** [comment]
@timholy No. A write barrier is needed when you assign an young and non-permanent object to the field of a old object so the few cases when this is not necessary are,

1, if the value assigned is known to be old or permanently rooted somewhere else e.g. `jl_nothing`, `jl_false` etc.
2. if the slot assigned to belongs to a known young object, the only case this is guaranteed is when the object is just allocated (i.e. it haven't go through the GC after a `gc_alloc` function). This is why no write barrier is needed for the code you linked.

**Comment by yuyichao** [comment]
Oh, of course another case where a write barrier is not necessary (also the only case where LLVM can perform the optimization automatically AFAIK) is if you are assigning an object to itself....

**Comment by timholy** [comment]
Thanks for explaining, @yuyichao. Too bad it wasn't a simple way to detect what caused that segfault.

**Comment by martinholters** [comment]
Anything preventing a merge here?

**Pull request merged**